File: compat/suites/dotnet-sdk/Groups/SqsGroup.cs
using Amazon.SQS.Model;
using OvercastCompat.Clients;
using OvercastCompat.Harness;

namespace OvercastCompat.Groups;

public sealed class SqsGroup(AwsClients clients) : IServiceGroup
{
    public IReadOnlyDictionary<string, TestFn> Impls() => new Dictionary<string, TestFn>(StringComparer.Ordinal)
    {
        ["CreateQueue"] = CreateQueueAsync,
        ["GetQueueUrl"] = GetQueueUrlAsync,
        ["ListQueues"] = ListQueuesAsync,
        ["SetQueueAttributes"] = SetQueueAttributesAsync,
        ["GetQueueAttributes"] = GetQueueAttributesAsync,
        ["TagQueue"] = TagQueueAsync,
        ["UntagQueue"] = UntagQueueAsync,
        ["DeleteQueue"] = DeleteQueueAsync,
        ["SendMessage"] = SendMessageAsync,
        ["SendMessageBatch"] = SendMessageBatchAsync,
        ["ReceiveMessage"] = ReceiveMessageAsync,
        ["DeleteMessage"] = DeleteMessageAsync,
        ["ChangeMessageVisibility"] = ChangeMessageVisibilityAsync,
        ["DeleteMessageBatch"] = DeleteMessageBatchAsync,
        ["PurgeQueue"] = PurgeQueueAsync,
        ["CreateDLQ"] = CreateDLQAsync,
        ["SetRedrivePolicy"] = SetRedrivePolicyAsync,
        ["GetRedrivePolicy"] = GetRedrivePolicyAsync,
        ["CreateFifoQueue"] = CreateFifoQueueAsync,
        ["SendFifoMessage"] = SendFifoMessageAsync,
        ["ReceiveFifoMessage"] = ReceiveFifoMessageAsync,
    };

    public IReadOnlyDictionary<string, SetupFn> Setups() => new Dictionary<string, SetupFn>(StringComparer.Ordinal)
    {
        ["sqs-queues"] = SetupQueuesAsync,
        ["sqs-messages"] = SetupMessagesAsync,
        ["sqs-dlq"] = SetupDlqAsync,
    };

    public IReadOnlyDictionary<string, SetupFn> Teardowns() => new Dictionary<string, SetupFn>(StringComparer.Ordinal)
    {
        ["sqs-queues"] = TeardownQueuesAsync,
        ["sqs-messages"] = TeardownMessagesAsync,
        ["sqs-dlq"] = TeardownDlqAsync,
        ["sqs-fifo"] = TeardownFifoAsync,
    };

    // ── sqs-queues ──

    private async Task SetupQueuesAsync(TestContext context)
    {
        var name = $"{context.RunId}-sqs-q";
        var response = await clients.SQS().CreateQueueAsync(new CreateQueueRequest { QueueName = name });
        context.Set("sqsQueueUrl", response.QueueUrl);
    }

    private async Task CreateQueueAsync(TestContext context)
    {
        var name = $"{context.RunId}-sqs-create-q";
        var response = await clients.SQS().CreateQueueAsync(new CreateQueueRequest { QueueName = name });
        var url = response.QueueUrl;
        Assertions.NotBlank(url, "CreateQueue: url");
        try
        {
            var list = await clients.SQS().ListQueuesAsync(new ListQueuesRequest { QueueNamePrefix = name });
            Assertions.True(list.QueueUrls.Any(u => u == url), $"CreateQueue: queue {name} not found in ListQueues (runId={context.RunId})");
        }
        finally
        {
            try { await clients.SQS().DeleteQueueAsync(new DeleteQueueRequest { QueueUrl = url }); } catch { }
        }
    }

    private async Task GetQueueUrlAsync(TestContext context)
    {
        var name = $"{context.RunId}-sqs-q";
        var response = await clients.SQS().GetQueueUrlAsync(new GetQueueUrlRequest { QueueName = name });
        Assertions.NotBlank(response.QueueUrl, "GetQueueUrl: url");
    }

    private async Task ListQueuesAsync(TestContext context)
    {
        var storedUrl = RequireQueueUrl(context, "sqsQueueUrl");
        var response = await clients.SQS().ListQueuesAsync(new ListQueuesRequest { QueueNamePrefix = $"{context.RunId}-sqs-q" });
        Assertions.True(response.QueueUrls.Any(u => u == storedUrl), $"ListQueues: queue URL not found (runId={context.RunId})");
    }

    private async Task SetQueueAttributesAsync(TestContext context)
    {
        var url = RequireQueueUrl(context, "sqsQueueUrl");
        await clients.SQS().SetQueueAttributesAsync(new SetQueueAttributesRequest
        {
            QueueUrl = url,
            Attributes = new Dictionary<string, string> { ["VisibilityTimeout"] = "120" },
        });
        var attrs = await clients.SQS().GetQueueAttributesAsync(new GetQueueAttributesRequest
        {
            QueueUrl = url,
            AttributeNames = ["VisibilityTimeout"],
        });
        Assertions.Equal("120", attrs.Attributes["VisibilityTimeout"], "SetQueueAttributes: VisibilityTimeout");
    }

    private async Task GetQueueAttributesAsync(TestContext context)
    {
        var url = RequireQueueUrl(context, "sqsQueueUrl");
        var response = await clients.SQS().GetQueueAttributesAsync(new GetQueueAttributesRequest
        {
            QueueUrl = url,
            AttributeNames = ["All"],
        });
        Assertions.True(response.Attributes.Count > 0, "GetQueueAttributes: expected non-empty attributes");
        Assertions.Equal("120", response.Attributes["VisibilityTimeout"], "GetQueueAttributes: VisibilityTimeout");
    }

    private async Task TagQueueAsync(TestContext context)
    {
        var url = RequireQueueUrl(context, "sqsQueueUrl");
        await clients.SQS().TagQueueAsync(new TagQueueRequest
        {
            QueueUrl = url,
            Tags = new Dictionary<string, string> { ["Environment"] = "compat", ["Team"] = "dotnet" },
        });
        var tags = await clients.SQS().ListQueueTagsAsync(new ListQueueTagsRequest { QueueUrl = url });
        Assertions.Equal("compat", tags.Tags["Environment"], "TagQueue: Environment tag");
    }

    private async Task UntagQueueAsync(TestContext context)
    {
        var url = RequireQueueUrl(context, "sqsQueueUrl");
        await clients.SQS().UntagQueueAsync(new UntagQueueRequest
        {
            QueueUrl = url,
            TagKeys = ["Environment", "Team"],
        });
        var tags = await clients.SQS().ListQueueTagsAsync(new ListQueueTagsRequest { QueueUrl = url });
        Assertions.True(tags.Tags.Count == 0, "UntagQueue: expected empty tags after untag");
    }

    private async Task DeleteQueueAsync(TestContext context)
    {
        var name = $"{context.RunId}-sqs-del-q";
        var create = await clients.SQS().CreateQueueAsync(new CreateQueueRequest { QueueName = name });
        var url = create.QueueUrl;
        await clients.SQS().DeleteQueueAsync(new DeleteQueueRequest { QueueUrl = url });
        var list = await clients.SQS().ListQueuesAsync(new ListQueuesRequest { QueueNamePrefix = name });
        Assertions.True(list.QueueUrls.Count == 0, $"DeleteQueue: queue {name} still present (runId={context.RunId})");
    }

    private async Task TeardownQueuesAsync(TestContext context)
    {
        var url = context.GetString("sqsQueueUrl");
        if (!string.IsNullOrWhiteSpace(url))
        {
            try { await clients.SQS().DeleteQueueAsync(new DeleteQueueRequest { QueueUrl = url }); } catch { }
        }
    }

    // ── sqs-messages ──

    private async Task SetupMessagesAsync(TestContext context)
    {
        var name = $"{context.RunId}-sqs-msg";
        var response = await clients.SQS().CreateQueueAsync(new CreateQueueRequest { QueueName = name });
        context.Set("sqsMsgUrl", response.QueueUrl);
    }

    private async Task SendMessageAsync(TestContext context)
    {
        var url = RequireQueueUrl(context, "sqsMsgUrl");
        var response = await clients.SQS().SendMessageAsync(new SendMessageRequest
        {
            QueueUrl = url,
            MessageBody = "hello from dotnet-sdk",
        });
        Assertions.NotBlank(response.MessageId, "SendMessage: MessageId");
    }

    private async Task SendMessageBatchAsync(TestContext context)
    {
        var url = RequireQueueUrl(context, "sqsMsgUrl");
        var response = await clients.SQS().SendMessageBatchAsync(new SendMessageBatchRequest
        {
            QueueUrl = url,
            Entries =
            [
                new SendMessageBatchRequestEntry { Id = "a", MessageBody = "batch-a" },
                new SendMessageBatchRequestEntry { Id = "b", MessageBody = "batch-b" },
            ],
        });
        Assertions.GreaterThanOrEqual(2, response.Successful.Count, "SendMessageBatch: expected >= 2 successful");
        Assertions.NotBlank(response.Successful[0].MessageId, "SendMessageBatch: MessageId[0]");
    }

    private async Task ReceiveMessageAsync(TestContext context)
    {
        var url = RequireQueueUrl(context, "sqsMsgUrl");
        await clients.SQS().SendMessageAsync(new SendMessageRequest { QueueUrl = url, MessageBody = "receive-test" });
        List<Message> messages = new();
        for (var i = 0; i < 5; i++)
        {
            var response = await clients.SQS().ReceiveMessageAsync(new ReceiveMessageRequest
            {
                QueueUrl = url,
                MaxNumberOfMessages = 10,
                WaitTimeSeconds = 1,
            });
            messages = response.Messages;
            if (messages.Count > 0)
            {
                break;
            }
        }
        Assertions.True(messages.Count > 0, "ReceiveMessage: expected at least 1 message");
    }

    private async Task DeleteMessageAsync(TestContext context)
    {
        var url = RequireQueueUrl(context, "sqsMsgUrl");
        await clients.SQS().SendMessageAsync(new SendMessageRequest { QueueUrl = url, MessageBody = "delete-me" });
        var recv = await clients.SQS().ReceiveMessageAsync(new ReceiveMessageRequest
        {
            QueueUrl = url,
            MaxNumberOfMessages = 1,
            WaitTimeSeconds = 3,
        });
        Assertions.True(recv.Messages.Count > 0, "DeleteMessage: expected at least 1 message");
        var handle = recv.Messages[0].ReceiptHandle;
        Assertions.NotBlank(handle, "DeleteMessage: ReceiptHandle");
        await clients.SQS().DeleteMessageAsync(new DeleteMessageRequest { QueueUrl = url, ReceiptHandle = handle });
    }

    private async Task ChangeMessageVisibilityAsync(TestContext context)
    {
        var url = RequireQueueUrl(context, "sqsMsgUrl");
        await clients.SQS().SendMessageAsync(new SendMessageRequest { QueueUrl = url, MessageBody = "visibility-test" });
        var recv = await clients.SQS().ReceiveMessageAsync(new ReceiveMessageRequest
        {
            QueueUrl = url,
            MaxNumberOfMessages = 1,
            WaitTimeSeconds = 3,
        });
        Assertions.True(recv.Messages.Count > 0, "ChangeMessageVisibility: expected at least 1 message");
        await clients.SQS().ChangeMessageVisibilityAsync(new ChangeMessageVisibilityRequest
        {
            QueueUrl = url,
            ReceiptHandle = recv.Messages[0].ReceiptHandle,
            VisibilityTimeout = 30,
        });
    }

    private async Task DeleteMessageBatchAsync(TestContext context)
    {
        var url = RequireQueueUrl(context, "sqsMsgUrl");
        await clients.SQS().SendMessageAsync(new SendMessageRequest { QueueUrl = url, MessageBody = "batch-delete-a" });
        await clients.SQS().SendMessageAsync(new SendMessageRequest { QueueUrl = url, MessageBody = "batch-delete-b" });
        var recv = await clients.SQS().ReceiveMessageAsync(new ReceiveMessageRequest
        {
            QueueUrl = url,
            MaxNumberOfMessages = 10,
            WaitTimeSeconds = 3,
        });
        Assertions.GreaterThanOrEqual(2, recv.Messages.Count, "DeleteMessageBatch: expected >= 2 messages");
        var batchResp = await clients.SQS().DeleteMessageBatchAsync(new DeleteMessageBatchRequest
        {
            QueueUrl = url,
            Entries = recv.Messages.Select((msg, idx) => new DeleteMessageBatchRequestEntry
            {
                Id = (idx + 1).ToString(),
                ReceiptHandle = msg.ReceiptHandle,
            }).ToList(),
        });
        Assertions.GreaterThanOrEqual(2, batchResp.Successful.Count, "DeleteMessageBatch: expected >= 2 successful deletes");
    }

    private async Task PurgeQueueAsync(TestContext context)
    {
        var url = RequireQueueUrl(context, "sqsMsgUrl");
        await clients.SQS().SendMessageAsync(new SendMessageRequest { QueueUrl = url, MessageBody = "purge-test" });
        await clients.SQS().PurgeQueueAsync(new PurgeQueueRequest { QueueUrl = url });
        var recv = await clients.SQS().ReceiveMessageAsync(new ReceiveMessageRequest
        {
            QueueUrl = url,
            MaxNumberOfMessages = 10,
            WaitTimeSeconds = 0,
        });
        Assertions.Equal(0, recv.Messages.Count, "PurgeQueue: expected no messages after purge");
    }

    private async Task TeardownMessagesAsync(TestContext context)
    {
        var url = context.GetString("sqsMsgUrl");
        if (!string.IsNullOrWhiteSpace(url))
        {
            try { await clients.SQS().DeleteQueueAsync(new DeleteQueueRequest { QueueUrl = url }); } catch { }
        }
    }

    // ── sqs-dlq ──

    private async Task SetupDlqAsync(TestContext context)
    {
        var srcName = $"{context.RunId}-sqs-src";
        var dlqName = $"{context.RunId}-sqs-dlq";
        var srcResponse = await clients.SQS().CreateQueueAsync(new CreateQueueRequest { QueueName = srcName });
        var dlqResponse = await clients.SQS().CreateQueueAsync(new CreateQueueRequest { QueueName = dlqName });
        context.Set("sqsSrcUrl", srcResponse.QueueUrl);
        context.Set("sqsDlqUrl", dlqResponse.QueueUrl);
    }

    private async Task CreateDLQAsync(TestContext context)
    {
        var srcUrl = context.GetString("sqsSrcUrl") ?? throw new InvalidOperationException("sqsSrcUrl not set");
        var dlqUrl = context.GetString("sqsDlqUrl") ?? throw new InvalidOperationException("sqsDlqUrl not set");
        var list = await clients.SQS().ListQueuesAsync(new ListQueuesRequest { QueueNamePrefix = $"{context.RunId}-sqs-src" });
        Assertions.True(list.QueueUrls.Any(u => u == srcUrl), $"CreateDLQ: src queue not found (runId={context.RunId})");
        var listDlq = await clients.SQS().ListQueuesAsync(new ListQueuesRequest { QueueNamePrefix = $"{context.RunId}-sqs-dlq" });
        Assertions.True(listDlq.QueueUrls.Any(u => u == dlqUrl), $"CreateDLQ: dlq queue not found (runId={context.RunId})");
    }

    private async Task SetRedrivePolicyAsync(TestContext context)
    {
        var srcUrl = context.GetString("sqsSrcUrl") ?? throw new InvalidOperationException("sqsSrcUrl not set");
        var dlqUrl = context.GetString("sqsDlqUrl") ?? throw new InvalidOperationException("sqsDlqUrl not set");
        var dlqAttrs = await clients.SQS().GetQueueAttributesAsync(new GetQueueAttributesRequest
        {
            QueueUrl = dlqUrl,
            AttributeNames = ["QueueArn"],
        });
        var dlqArn = dlqAttrs.Attributes["QueueArn"];
        var policy = $"{{\"maxReceiveCount\":\"3\",\"deadLetterTargetArn\":\"{dlqArn}\"}}";
        await clients.SQS().SetQueueAttributesAsync(new SetQueueAttributesRequest
        {
            QueueUrl = srcUrl,
            Attributes = new Dictionary<string, string> { ["RedrivePolicy"] = policy },
        });
        var srcAttrs = await clients.SQS().GetQueueAttributesAsync(new GetQueueAttributesRequest
        {
            QueueUrl = srcUrl,
            AttributeNames = ["RedrivePolicy"],
        });
        Assertions.True(srcAttrs.Attributes.ContainsKey("RedrivePolicy"), "SetRedrivePolicy: RedrivePolicy not found");
        Assertions.True(srcAttrs.Attributes["RedrivePolicy"].Contains("deadLetterTargetArn"), "SetRedrivePolicy: missing deadLetterTargetArn");
    }

    private async Task GetRedrivePolicyAsync(TestContext context)
    {
        var srcUrl = context.GetString("sqsSrcUrl") ?? throw new InvalidOperationException("sqsSrcUrl not set");
        var attrs = await clients.SQS().GetQueueAttributesAsync(new GetQueueAttributesRequest
        {
            QueueUrl = srcUrl,
            AttributeNames = ["RedrivePolicy"],
        });
        Assertions.True(attrs.Attributes.ContainsKey("RedrivePolicy"), "GetRedrivePolicy: RedrivePolicy not found");
        Assertions.True(attrs.Attributes["RedrivePolicy"].Contains("maxReceiveCount"), "GetRedrivePolicy: missing maxReceiveCount");
    }

    private async Task TeardownDlqAsync(TestContext context)
    {
        var srcUrl = context.GetString("sqsSrcUrl");
        if (!string.IsNullOrWhiteSpace(srcUrl))
        {
            try { await clients.SQS().DeleteQueueAsync(new DeleteQueueRequest { QueueUrl = srcUrl }); } catch { }
        }
        var dlqUrl = context.GetString("sqsDlqUrl");
        if (!string.IsNullOrWhiteSpace(dlqUrl))
        {
            try { await clients.SQS().DeleteQueueAsync(new DeleteQueueRequest { QueueUrl = dlqUrl }); } catch { }
        }
    }

    // ── sqs-fifo ──

    private async Task CreateFifoQueueAsync(TestContext context)
    {
        var name = $"{context.RunId}-sqs-fifo.fifo";
        var response = await clients.SQS().CreateQueueAsync(new CreateQueueRequest
        {
            QueueName = name,
            Attributes = new Dictionary<string, string> { ["FifoQueue"] = "true" },
        });
        var url = response.QueueUrl;
        Assertions.NotBlank(url, "CreateFifoQueue: url");
        context.Set("sqsFifoUrl", url);
    }

    private async Task SendFifoMessageAsync(TestContext context)
    {
        var url = context.GetString("sqsFifoUrl") ?? throw new InvalidOperationException("sqsFifoUrl not set");
        var response = await clients.SQS().SendMessageAsync(new SendMessageRequest
        {
            QueueUrl = url,
            MessageBody = "fifo-message",
            MessageGroupId = "test-group",
            MessageDeduplicationId = $"{context.RunId}-dedup-1",
        });
        Assertions.NotBlank(response.MessageId, "SendFifoMessage: MessageId");
    }

    private async Task ReceiveFifoMessageAsync(TestContext context)
    {
        var url = context.GetString("sqsFifoUrl") ?? throw new InvalidOperationException("sqsFifoUrl not set");
        List<Message> messages = new();
        for (var i = 0; i < 5; i++)
        {
            var response = await clients.SQS().ReceiveMessageAsync(new ReceiveMessageRequest
            {
                QueueUrl = url,
                MaxNumberOfMessages = 10,
                WaitTimeSeconds = 1,
            });
            messages = response.Messages;
            if (messages.Count > 0)
            {
                break;
            }
        }
        Assertions.True(messages.Count > 0, "ReceiveFifoMessage: expected at least 1 message");
    }

    private async Task TeardownFifoAsync(TestContext context)
    {
        var url = context.GetString("sqsFifoUrl");
        if (!string.IsNullOrWhiteSpace(url))
        {
            try { await clients.SQS().DeleteQueueAsync(new DeleteQueueRequest { QueueUrl = url }); } catch { }
        }
    }

    private static string RequireQueueUrl(TestContext context, string key)
    {
        return context.GetString(key) ?? throw new InvalidOperationException($"{key} not set");
    }
}

File: compat/suites/java-sdk/src/main/java/io/overcast/compat/groups/SqsGroup.java
package io.overcast.compat.groups;

import io.overcast.compat.clients.AwsClients;
import io.overcast.compat.harness.Assertions;
import io.overcast.compat.harness.TestContext;
import io.overcast.compat.harness.TestFn;
import software.amazon.awssdk.services.sqs.SqsClient;
import software.amazon.awssdk.services.sqs.model.*;

import java.util.List;
import java.util.Map;

/**
 * SQS compatibility test group.
 *
 * <p>Groups: sqs-queues, sqs-messages, sqs-dlq, sqs-fifo.
 */
public final class SqsGroup implements ServiceGroup {

    private final AwsClients clients;

    public SqsGroup(AwsClients clients) {
        this.clients = clients;
    }

    private SqsClient sqs() { return clients.sqs(); }

    @Override
    public Map<String, TestFn> impls() {
        return Map.ofEntries(
                Map.entry("CreateQueue",              this::createQueue),
                Map.entry("GetQueueUrl",              this::getQueueUrl),
                Map.entry("ListQueues",               this::listQueues),
                Map.entry("SetQueueAttributes",       this::setQueueAttributes),
                Map.entry("GetQueueAttributes",       this::getQueueAttributes),
                Map.entry("TagQueue",                 this::tagQueue),
                Map.entry("UntagQueue",               this::untagQueue),
                Map.entry("DeleteQueue",              this::deleteQueue),
                Map.entry("SendMessage",              this::sendMessage),
                Map.entry("SendMessageBatch",         this::sendMessageBatch),
                Map.entry("ReceiveMessage",           this::receiveMessage),
                Map.entry("DeleteMessage",            this::deleteMessage),
                Map.entry("ChangeMessageVisibility",  this::changeMessageVisibility),
                Map.entry("DeleteMessageBatch",       this::deleteMessageBatch),
                Map.entry("PurgeQueue",               this::purgeQueue),
                Map.entry("CreateDLQ",                this::createDlq),
                Map.entry("SetRedrivePolicy",         this::setRedrivePolicy),
                Map.entry("GetRedrivePolicy",         this::getRedrivePolicy),
                Map.entry("CreateFifoQueue",          this::createFifoQueue),
                Map.entry("SendFifoMessage",          this::sendFifoMessage),
                Map.entry("ReceiveFifoMessage",       this::receiveFifoMessage)
        );
    }

    @Override
    public Map<String, TestFn> setups() {
        return Map.ofEntries(
                Map.entry("sqs-queues",   this::setupQueues),
                Map.entry("sqs-messages", this::setupMessages),
                Map.entry("sqs-dlq",      this::setupDlq),
                Map.entry("sqs-fifo",     this::setupFifo)
        );
    }

    @Override
    public Map<String, TestFn> teardowns() {
        return Map.ofEntries(
                Map.entry("sqs-queues",   ctx -> deleteQueueSilently(ctx.getString("sqsQueueUrl"))),
                Map.entry("sqs-messages", ctx -> deleteQueueSilently(ctx.getString("sqsMsgQueueUrl"))),
                Map.entry("sqs-dlq",      this::teardownDlq),
                Map.entry("sqs-fifo",     ctx -> deleteQueueSilently(ctx.getString("sqsFifoUrl")))
        );
    }

    // ── sqs-queues ─────────────────────────────────────────────────────────────

    private void setupQueues(TestContext ctx) throws Exception {
        String name = ctx.runId() + "-sqsq";
        var resp = sqs().createQueue(r -> r.queueName(name));
        ctx.set("sqsQueueUrl", resp.queueUrl());
        ctx.set("sqsQueueName", name);
    }

    private void createQueue(TestContext ctx) {
        // Queue created in setup — verify URL is set.
        Assertions.assertNotBlank(ctx.getString("sqsQueueUrl"), "sqsQueueUrl");
    }

    private void getQueueUrl(TestContext ctx) throws Exception {
        String name = ctx.getString("sqsQueueName");
        var resp = sqs().getQueueUrl(r -> r.queueName(name));
        Assertions.assertNotBlank(resp.queueUrl(), "GetQueueUrl: queueUrl");
    }

    private void listQueues(TestContext ctx) throws Exception {
        String name = ctx.getString("sqsQueueName");
        var resp = sqs().listQueues(r -> r.queueNamePrefix(name));
        boolean found = resp.queueUrls().stream().anyMatch(u -> u.contains(name));
        Assertions.assertTrue(found, "ListQueues: queue " + name + " not found");
    }

    private void setQueueAttributes(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsQueueUrl");
        sqs().setQueueAttributes(r -> r.queueUrl(url)
                .attributes(Map.of(QueueAttributeName.VISIBILITY_TIMEOUT, "60")));
    }

    private void getQueueAttributes(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsQueueUrl");
        var resp = sqs().getQueueAttributes(r -> r.queueUrl(url)
                .attributeNames(QueueAttributeName.ALL));
        Assertions.assertTrue(resp.hasAttributes(), "GetQueueAttributes: no attributes returned");
        String vt = resp.attributesAsStrings().get("VisibilityTimeout");
        Assertions.assertEquals("60", vt, "GetQueueAttributes: VisibilityTimeout mismatch");
    }

    private void tagQueue(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsQueueUrl");
        sqs().tagQueue(r -> r.queueUrl(url).tags(Map.of("env", "test")));
    }

    private void untagQueue(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsQueueUrl");
        sqs().untagQueue(r -> r.queueUrl(url).tagKeys("env"));
    }

    private void deleteQueue(TestContext ctx) throws Exception {
        // The teardown deletes the setup queue; this test creates an ephemeral one.
        String name = ctx.runId() + "-sqsdel";
        var resp = sqs().createQueue(r -> r.queueName(name));
        sqs().deleteQueue(r -> r.queueUrl(resp.queueUrl()));
    }

    // ── sqs-messages ──────────────────────────────────────────────────────────

    private void setupMessages(TestContext ctx) throws Exception {
        String name = ctx.runId() + "-sqsmsg";
        var resp = sqs().createQueue(r -> r.queueName(name));
        ctx.set("sqsMsgQueueUrl", resp.queueUrl());
    }

    private void sendMessage(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsMsgQueueUrl");
        var resp = sqs().sendMessage(r -> r.queueUrl(url).messageBody("hello SQS"));
        Assertions.assertNotBlank(resp.messageId(), "SendMessage: messageId");
    }

    private void sendMessageBatch(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsMsgQueueUrl");
        var resp = sqs().sendMessageBatch(r -> r.queueUrl(url).entries(
                SendMessageBatchRequestEntry.builder().id("m1").messageBody("batch-1").build(),
                SendMessageBatchRequestEntry.builder().id("m2").messageBody("batch-2").build()
        ));
        Assertions.assertEquals(2, resp.successful().size(), "SendMessageBatch: expected 2 successful");
    }

    private void receiveMessage(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsMsgQueueUrl");
        var resp = sqs().receiveMessage(r -> r.queueUrl(url).maxNumberOfMessages(10).waitTimeSeconds(0));
        Assertions.assertNotEmpty(resp.messages(), "ReceiveMessage: no messages received");
        ctx.set("sqsReceiptHandle", resp.messages().get(0).receiptHandle());
    }

    private void deleteMessage(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsMsgQueueUrl");
        String rh  = ctx.getString("sqsReceiptHandle");
        if (rh != null) {
            sqs().deleteMessage(r -> r.queueUrl(url).receiptHandle(rh));
        }
    }

    private void changeMessageVisibility(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsMsgQueueUrl");
        // Send a fresh message to change visibility on.
        sqs().sendMessage(r -> r.queueUrl(url).messageBody("vis-test"));
        var recv = sqs().receiveMessage(r -> r.queueUrl(url).maxNumberOfMessages(1).waitTimeSeconds(0));
        if (recv.messages().isEmpty()) return; // skip gracefully if queue is empty
        String rh = recv.messages().get(0).receiptHandle();
        sqs().changeMessageVisibility(r -> r.queueUrl(url).receiptHandle(rh).visibilityTimeout(30));
    }

    private void deleteMessageBatch(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsMsgQueueUrl");
        sqs().sendMessage(r -> r.queueUrl(url).messageBody("batch-del-1"));
        sqs().sendMessage(r -> r.queueUrl(url).messageBody("batch-del-2"));
        var recv = sqs().receiveMessage(r -> r.queueUrl(url).maxNumberOfMessages(10).waitTimeSeconds(0));
        if (recv.messages().isEmpty()) return;
        var entries = recv.messages().stream()
                .map(m -> DeleteMessageBatchRequestEntry.builder()
                        .id(m.messageId()).receiptHandle(m.receiptHandle()).build())
                .toList();
        sqs().deleteMessageBatch(r -> r.queueUrl(url).entries(entries));
    }

    private void purgeQueue(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsMsgQueueUrl");
        sqs().purgeQueue(r -> r.queueUrl(url));
    }

    // ── sqs-dlq ────────────────────────────────────────────────────────────────

    private void setupDlq(TestContext ctx) throws Exception {
        String dlqName  = ctx.runId() + "-sqsdlq";
        String srcName  = ctx.runId() + "-sqssrc";
        var dlqResp     = sqs().createQueue(r -> r.queueName(dlqName));
        var dlqAttrs    = sqs().getQueueAttributes(r -> r
                .queueUrl(dlqResp.queueUrl())
                .attributeNames(QueueAttributeName.QUEUE_ARN));
        String dlqArn   = dlqAttrs.attributesAsStrings().get("QueueArn");
        var srcResp     = sqs().createQueue(r -> r.queueName(srcName));
        ctx.set("sqsDlqUrl", dlqResp.queueUrl());
        ctx.set("sqsDlqArn", dlqArn);
        ctx.set("sqsSrcUrl", srcResp.queueUrl());
    }

    private void teardownDlq(TestContext ctx) {
        deleteQueueSilently(ctx.getString("sqsDlqUrl"));
        deleteQueueSilently(ctx.getString("sqsSrcUrl"));
    }

    private void createDlq(TestContext ctx) {
        Assertions.assertNotBlank(ctx.getString("sqsDlqUrl"), "sqsDlqUrl");
    }

    private void setRedrivePolicy(TestContext ctx) throws Exception {
        String url    = ctx.getString("sqsSrcUrl");
        String dlqArn = ctx.getString("sqsDlqArn");
        String policy = "{\"deadLetterTargetArn\":\"" + dlqArn + "\",\"maxReceiveCount\":\"3\"}";
        sqs().setQueueAttributes(r -> r.queueUrl(url)
                .attributes(Map.of(QueueAttributeName.REDRIVE_POLICY, policy)));
    }

    private void getRedrivePolicy(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsSrcUrl");
        var resp = sqs().getQueueAttributes(r -> r
                .queueUrl(url).attributeNames(QueueAttributeName.REDRIVE_POLICY));
        Assertions.assertTrue(resp.attributesAsStrings().containsKey("RedrivePolicy"),
                "GetRedrivePolicy: RedrivePolicy attribute not present");
    }

    // ── sqs-fifo ───────────────────────────────────────────────────────────────

    private void setupFifo(TestContext ctx) throws Exception {
        String name = ctx.runId() + "-sqsfifo.fifo";
        var resp = sqs().createQueue(r -> r.queueName(name)
                .attributes(Map.of(QueueAttributeName.FIFO_QUEUE, "true")));
        ctx.set("sqsFifoUrl", resp.queueUrl());
    }

    private void createFifoQueue(TestContext ctx) {
        Assertions.assertNotBlank(ctx.getString("sqsFifoUrl"), "sqsFifoUrl");
    }

    private void sendFifoMessage(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsFifoUrl");
        var resp = sqs().sendMessage(r -> r.queueUrl(url)
                .messageBody("fifo-msg")
                .messageGroupId("grp1")
                .messageDeduplicationId("dedup1"));
        Assertions.assertNotBlank(resp.messageId(), "SendFifoMessage: messageId");
    }

    private void receiveFifoMessage(TestContext ctx) throws Exception {
        String url = ctx.getString("sqsFifoUrl");
        var resp = sqs().receiveMessage(r -> r.queueUrl(url)
                .maxNumberOfMessages(1).waitTimeSeconds(2));
        Assertions.assertNotEmpty(resp.messages(), "ReceiveFifoMessage: no messages received");
        Assertions.assertEquals("fifo-msg", resp.messages().get(0).body(),
                "ReceiveFifoMessage: body mismatch");
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private void deleteQueueSilently(String url) {
        if (url == null) return;
        try { sqs().deleteQueue(r -> r.queueUrl(url)); } catch (Exception ignored) {}
    }
}
